# ML Zoomcamp: Homework 1

# Q1. Pandas version

What version of Pandas did you install?

You can get the version information using the __version__ field:

In [4]:
import numpy as np
import pandas as pd

pd.__version__ #Q1


'3.0.5'

# Q2. Records count
How many records are in the dataset?

In [16]:
df = pd.read_csv("car_fuel_efficiency_2026.csv")
df.shape  # Q2: records count


(10000, 11)

There are 10 000 records in the dataset.

# Q3. Fuel types
How many fuel types are presented in the dataset?

In [19]:
fuel_types = df['fuel_type'].unique()
print(fuel_types)

<StringArray>
['Gasoline', 'Diesel', 'Hybrid']
Length: 3, dtype: str


There are 3 unique fuel types

## Q 4. Missing values

How many columns in the dataset have missing values?

In [21]:
df.isnull().sum()

model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors                0
engine_displacement      0
num_cylinders            0
horsepower             877
vehicle_weight           0
acceleration           264
fuel_efficiency_mpg      0
dtype: int64

Two coulmns have missing values

## Q5. Max fuel efficiency
What's the maximum fuel efficiency of cars from Asia?

In [26]:
max_fuel_efficiency_asia = df[df['origin'] == 'Asia']['fuel_efficiency_mpg'].max()
print(max_fuel_efficiency_asia)

41.2


Max fuel efficiency of cars in Asia is 41.2 

# Q6. Median value of horsepower
* Find the median value of the horsepower column in the dataset.
* Next, calculate the most frequent value of the same horsepower column.
*Use the fillna method to fill the missing values in the horsepower column with the most frequent value from the previous step.
* Now, calculate the median value of horsepower once again.

In [30]:
hp = df['horsepower'].median()
print(hp)

254.0


In [35]:
hp_mode  = df['horsepower'].mode()
print(hp_mode)

0    252.0
Name: horsepower, dtype: float64


In [36]:
hp_mode_value = hp_mode[0]
print(hp_mode_value)

252.0


In [37]:
horsepower_filled = df['horsepower'].fillna(hp_mode_value)

In [39]:
median_horsepower = horsepower_filled.median()
print(median_horsepower)


252.0


Has it changed?

The median decreased after filling in  Nan values. 

# Q7. Sum of weights


* Select all the cars from Asia
* Select only columns vehicle_weight and model_year
* Select the first 7 values
* Get the underlying NumPy array. Let's call it X.
* Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. * Let's call the result XTX.
* Invert XTX.
* Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200].
* Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
* What's the sum of all the elements of the result?

In [42]:
asian_cars = df[df['origin'] == 'Asia']

In [44]:
asian_subset = asian_cars[[ 'vehicle_weight' , 'model_year']]

In [47]:
first_seven = asian_subset.head(7)
print(first_seven)

    vehicle_weight  model_year
2             4240        1996
6             4310        1992
8             4230        1997
10            4420        1978
15            3950        2007
16            4750        2021
22            4450        2019


In [48]:
X = first_seven.values
print(X)

[[4240 1996]
 [4310 1992]
 [4230 1997]
 [4420 1978]
 [3950 2007]
 [4750 2021]
 [4450 2019]]


In [52]:
XTX = X.T @ X
print(XTX)

[[131950500  60750580]
 [ 60750580  28041424]]


In [54]:
XTX_inv = np.linalg.inv(XTX)
print(XTX_inv)

[[ 2.96830537e-06 -6.43071025e-06]
 [-6.43071025e-06  1.39675281e-05]]


In [55]:
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])

In [57]:
w = XTX_inv @ X.T @ y

In [58]:
print(w.sum()) # the sum of all the elements of the result

0.36919696904927535
